Feature 10

fjank opened the pull request:
resolves #10

fjank committed:
First BDD test added, resolves #10

Do note that there are edge-cases not handled in this commit, as a test
for the code-reviewer. :)

fjank committed:
First BDD test added

Do note that there are edge-cases not handled in this commit, as a test
for the code-reviewer. :)

fjank committed:
Merge branch 'feature-10' of github.com:fjank/github-workflow-kata into feature-10

netbrain reviewed (dismissed):
Nothing wrong with the code, other than it feels a bit over-engineered for the first issue. I would have preferred a simpler implementation, the least complex solution to make the test pass, and evolve the code as new requirements are added to the codebase. But that being said, it is quite typical for developers to get ahead of themselves :) me included!

Basically letting the tests drive the architecture of the code with a focus of only solving the given task at hand at any point in time.

As for cucumber, it's basically a way of giving business people a way to define tests and letting developers implement them. If i'm to say anything about cucumber it would be that I'd prefer a way to test grid's more easily. As you probably want it to be both writable and readable for the business people. An example would be:

```
Given the grid:
  *..
  ...
  ...

Then the next generation should look like:
  ...
  ...
  ...                        

```

However there might be technical limitations to cucumber which doesn't make this possible, idk.

fjank commented:
Agree to both comments, will restructure the cucumber tests, and massage the code to make it the least complex and only keep code to make the tests green. Pure BDD/TDD driven code!

netbrain commented:
As an example. Say you were working on #4 instead. And say you write the tests first, then given the test (pseudocode):

```
testShouldGiveTextualRepresentationOfGrid(){
 Grid g = new Grid(3,3);
 assertEquals(
    "...\n"+
    "...\n"+
    "...\n",
    g.toString()
  );
}
```

Now for the kicker, the following is a perfectly fine implementation for this first test:

```
class Grid {
  public Grid(int w, int h){}
  public String toString(){
    return "...\n...\n...\n";
  }
}
```

So as you add business logic you should try to solve the tests as simple as possible and force yourself to create tests that ultimately enforces correct behavior.

So for the next test you would probably try to create a grid with a different size, say 4 by 4. Which in turn would have to actually store that information and depending on that output the correct string for 3x3 and 4x4. after that you would probably add notation for living cells. 

It's like you'r trying to play two roles, the tester and the implementor, where the implementor is constantly trying to trick the tester and create silly logic that actually wouldn't hold in real life. And then the tester needs to counteract those implementations with more and more elaborate tests.

Now ofc. in real life you probably don't do this even as a seasoned TDD practioner, however I still think it's useful to do this in honing one's skills. It forces you to think deeply on the current situation.

fjank committed:
Simplified solution, and enhanced cucumber spec to be more usable.

As suggested I have now simplified the solution:
- There is now only a Grid class with all functionality.
- Grid class implements equals and hashCode to make comparisons between
grids simple.
- toString for grids displays the complete grid, to make it easier to
debug when results are not correct.
- the cucumber features now support defining a grid "initial state" and
an expected state.

fjank commented:
Gah! I went to far again, I see where you want to go. I will add another commit that I think will represent this even better, removing most of the scenarios (leaving only one), and the absolutely simplest code to make it work.

fjank committed:
Super-simplified the feature.

It is now supersimplified, and the single cucumber  feature works as
expected.

netbrain commented on src/test/resources/features/two-neighbour-rule.feature:
```diff
@@ -0,0 +1,19 @@
+      .*.
+      ...
+    """
+    When the generation is incremented
```
```suggestion
```
Can leave it out. The "Then" should handle generation incrementation when you get to that

fjank committed:
Update src/test/resources/features/two-neighbour-rule.feature

Co-authored-by: Kim Eik <kim@heldig.org>

fjank commented on src/test/java/features/TwoNeighbourRuleSteps.java:
```diff
@@ -0,0 +1,30 @@
+            grid = new Grid();
+        });
+
+        When("the generation is incremented", () -> {
+
+        });
```
```suggestion

```

fjank reviewed (commented):
should also remove the implementation for the When step.

fjank committed:
Update src/test/java/features/TwoNeighbourRuleSteps.java

fjank committed:
A bit to simple, the test did not use the input

The test would always be green (faulty test)
extended test to check grid using the input from the features,
thus extending grid to accept grid input string in constructor.
